test/test-mc-fle2-payload-ieev.c

```diff
@@ -0,0 +1,284 @@
+   _mongocrypt_buffer_copy_from_hex (&expect_K_KeyId,
+                                     "abcdefab123498761234123456789012");
+   _mongocrypt_buffer_copy_from_hex (
+      &correct_S_Key,
+      "7dbfebc619aa68a659f64b8e23ccd21644ac326cb74a26840c3d2420176c40ae088294d0"
+      "0ad6cae9684237b21b754cf503f085c25cd320bf035c3417416e1e6fe3d9219f79586582"
+      "112740b2add88e1030d91926ae8afc13ee575cfb8bb965b7");
+   _mongocrypt_buffer_copy_from_hex (
+      &correct_K_Key,
+      "a7ddbc4c8be00d51f68d9d8e485f351c8edc8d2206b24d8e0e1816d005fbe520e4891250"
+      "47d647b0d8684bfbdbf09c304085ed086aba6c2b2b1677ccc91ced8847a733bf5e5682c8"
+      "4b3ee7969e4a5fe0e0c21e5e3ee190595a55f83147d8de2a");
+   ASSERT (_mongocrypt_buffer_copy_from_data_and_size (
+      &expect_client_value,
+      (const uint8_t *) "\x09\x00\x00\x00value123\x00",
+      13));
+
+   /* Test success. */
+   {
+      mongocrypt_status_t *status = mongocrypt_status_new ();
+      ieev = mc_FLE2IndexedEqualityEncryptedValue_new ();
+      ASSERT_OK_STATUS (
+         mc_FLE2IndexedEqualityEncryptedValue_parse (ieev, &input, status),
+         status);
+
+      const _mongocrypt_buffer_t *got =
+         mc_FLE2IndexedEqualityEncryptedValue_get_S_KeyId (ieev, status);
+      ASSERT_OR_PRINT (got != NULL, status);
+      ASSERT_CMPBUF (expect_S_KeyId, *got);
+
+      ASSERT_OK_STATUS (mc_FLE2IndexedEqualityEncryptedValue_add_S_Key (
+                           crypt->crypto, ieev, &correct_S_Key, status),
+                        status);
+
+      got = mc_FLE2IndexedEqualityEncryptedValue_get_K_KeyId (ieev, status);
+      ASSERT_OR_PRINT (got != NULL, status);
+      ASSERT_CMPBUF (expect_K_KeyId, *got);
+
+      ASSERT_OK_STATUS (mc_FLE2IndexedEqualityEncryptedValue_add_K_Key (
+                           crypt->crypto, ieev, &correct_K_Key, status),
+                        status);
+      got = mc_FLE2IndexedEqualityEncryptedValue_get_ClientValue (ieev, status);
+      ASSERT_CMPBUF (expect_client_value, *got);
+      mc_FLE2IndexedEqualityEncryptedValue_destroy (ieev);
+      mongocrypt_status_destroy (status);
+   }
+
+   /* Test an incorrect S_Key. */
+   {
+      mongocrypt_status_t *status = mongocrypt_status_new ();
+      _mongocrypt_buffer_t incorrect_S_Key;
+
+      _mongocrypt_buffer_init (&incorrect_S_Key);
+      _mongocrypt_buffer_copy_to (&correct_S_Key, &incorrect_S_Key);
+      /* The last 32 bytes of S_Key are used to generate
+       * ServerDataEncryptionLevel1Token. Change last byte to make S_Key
+       * incorrect. */
+      incorrect_S_Key.data[incorrect_S_Key.len - 1] = 0;
+      ieev = mc_FLE2IndexedEqualityEncryptedValue_new ();
+      ASSERT_OK_STATUS (
+         mc_FLE2IndexedEqualityEncryptedValue_parse (ieev, &input, status),
+         status);
+      /* Since S_Key is used for non-AEAD encryption. Decryption does not return
+       * an error. The output is garbled. It fails to parse the decrypted Inner
+       * struct. */
```

Comment:
```suggestion
      /* Since S_Key is used for non-AEAD encryption, decryption does not return
       * an error. The output is garbled. It fails to parse the decrypted Inner
       * struct. */
```